### PR TITLE
Align the logic we have for document syncing with project syncing

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 1]
```diff
@@ -482,22 +535,6 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
                     project = addDocuments(project.Solution, lazyDocumentsToAdd.ToImmutable()).GetProject(project.Id)!;
                 }
 
-                // changed document
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:485]: moved 'changed' to the bottom, after we process adds/removes (matches what we do with projects).

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 2]
```diff
@@ -58,13 +58,14 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
...
 
-                    var oldSolutionCompilationChecksums = await solution.CompilationState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
                     var newSolutionCompilationChecksums = await _assetProvider.GetAssetAsync<SolutionCompilationStateChecksums>(
                         assetHint: AssetHint.None, newSolutionChecksum, cancellationToken).ConfigureAwait(false);
-                    var oldSolutionChecksums = await solution.CompilationState.SolutionState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
                     var newSolutionChecksums = await _assetProvider.GetAssetAsync<SolutionStateChecksums>(
                         assetHint: AssetHint.None, newSolutionCompilationChecksums.SolutionState, cancellationToken).ConfigureAwait(false);
 
+                    var oldSolutionCompilationChecksums = await solution.CompilationState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
+                    var oldSolutionChecksums = await solution.CompilationState.SolutionState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:67]: same as before.  i just think it's clearer to do all the new checksum fetching prior to computing the local checksums.

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 3]
```diff
@@ -438,21 +440,61 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
                 Func<Solution, ImmutableArray<DocumentId>, Solution> removeDocuments,
                 CancellationToken cancellationToken) where TDocumentState : TextDocumentState
             {
-                using var _1 = PooledHashSet<Checksum>.GetInstance(out var olds);
-                using var _2 = PooledHashSet<Checksum>.GetInstance(out var news);
+                using var _1 = PooledDictionary<DocumentId, Checksum>.GetInstance(out var oldDocumentIdToChecksum);
+                using var _2 = PooledDictionary<DocumentId, Checksum>.GetInstance(out var newDocumentIdToChecksum);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:444]: aligned with projects, and cleaner as dictionaries for the processing we need to do (we want the doc-ids, and htis preserves tehm).

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 4]
```diff
@@ -438,21 +440,61 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
...
+                    oldDocumentIdToChecksum.Add(documentId, oldChecksum);
+
+                foreach (var (newChecksum, documentId) in newChecksums)
+                    newDocumentIdToChecksum.Add(documentId, newChecksum);
+
+                allDocumentIds.AddRange(oldChecksums.Ids);
+                allDocumentIds.AddRange(newChecksums.Ids);
+
+                // remove documents that are the same on both sides
+                // remove projects that are the same on both sides.
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:457]:
```suggestion
```

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 5]
```diff
@@ -462,6 +503,17 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
                     await _assetProvider.SynchronizeProjectAssetsAsync(projectChecksums, cancellationToken).ConfigureAwait(false);
                 }
 
+                return await UpdateDocumentsAsync(project, addDocuments, removeDocuments, oldDocumentIdToStateChecksums, newDocumentIdToStateChecksums, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:506]: extracted into helper (like we do with projects).  helps keep the syncing of checksums distinct from teh updating of state.

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 6]
```diff
@@ -438,21 +440,60 @@ private async Task<Project> UpdateProjectInfoAsync(Project project, Checksum inf
...
-                olds.AddRange(oldChecksums.Checksums.Children);
-                news.AddRange(newChecksums.Checksums.Children);
+                allDocumentIds.AddRange(oldChecksums.Ids);
+                allDocumentIds.AddRange(newChecksums.Ids);
 
-                // remove documents that exist in both side
-                olds.ExceptWith(newChecksums.Checksums);
-                news.ExceptWith(oldChecksums.Checksums);
+                // remove documents that are the same on both sides
+                foreach (var documentId in allDocumentIds)
```
Comment (ToddGrun) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:457]: >foreach (var documentId in allDocumentIds) [](http://example.com/codeflow?start=16&length=42)

Can you just enumerate over oldChecksums and not need allDocumentIds?

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs]: yup.  done.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**